### PR TITLE
[MRG] FIX/MAINT/ENH Covlove clip rank

### PR DESCRIPTION
**Pull request by dengemann**
Work in progress & to be fine tuned.

But here is what I propose (pooling over the last months of discussion):

1) detect the rank prior to cov estimation and clip the cov if the rank is < n_sensors. 
This should then stabilize the performance of plot_white
2) compute SSSed covariance in a low dimensional space to save time and improve regularization.

cc @agramfort @larsoner

### CODE REVIEW

**Comment by larsoner** [comment]
These changes look too extensive for 0.15. I think we'll want to test for some months before unleashing on the public. 0.16 okay with you?

**Comment by agramfort** [comment]
@larsoner give us one more day to evaluate.

It's more bug fixes

File: mne/io/tests/test_proc_history.py [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 import numpy as np
 import os.path as op
-from mne.io import read_info
+from mne.io import read_info, read_raw_fif
 from mne.io.constants import FIFF
-from mne.io.proc_history import _get_sss_rank
+from mne import _get_rank_sss
```
Comment (larsoner) [inline mne/io/tests/test_proc_history.py]: I don't think we should put this at the root level, it seems better buried in `io.proc_history`

**Comment by dengemann** [comment]
I think this solves a circrular import issue. I could not get it working
inside meas info IIRC ...

**Comment by larsoner** [comment]
> I think this solves a circrular import issue.

Probably better to solve it with a nested import

**Comment by dengemann** [comment]
Let me try that. I agree that it is far from ideal as it is now.

File: tutorials/plot_compute_covariance.py [hunk 2]
```diff
@@ -3,29 +3,40 @@
 
 Computing covariance matrix
 ===========================
-
-Many methods in MNE, including e.g. source estimation and some classification
```
Comment (larsoner) [inline tutorials/plot_compute_covariance.py]: This broke CircleCI

```
ValueError: Example docstring should have a header for the example title and at least a paragraph explaining what the example is about. Please check the example file:
 plot_compute_covariance.py
```

Comment (dengemann) [reply tutorials/plot_compute_covariance.py]: wow, oops. this was not intended.

**Comment by dengemann** [comment]
Circle CI is happy: https://6087-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.Vslzm8M/html/auto_tutorials/plot_compute_covariance.html#sphx-glr-auto-tutorials-plot-compute-covariance-py

**Comment by dengemann** [comment]
@larsoner @larsoner @jasmainak this is good for review.

File: mne/cov.py [hunk 3]
```diff
@@ -854,6 +855,91 @@ def _unpack_epochs(epochs):
...
+    has_sss = False
+    if len(info['proc_history']) > 0:
+        has_sss = (info['proc_history'][0].get('max_info') is not None)
+    if has_sss:
+        picks_list_ = _merge_picks_list(picks_list)
+    else:
+        picks_list_ = picks_list
+
+    for key, this_picks in picks_list_:
+        this_info = pick_info(info.copy(), this_picks)
```
Comment (larsoner) [inline mne/cov.py]: `pick_info` has a `copy` param and I think it defaults to `True`

File: mne/cov.py [hunk 4]
```diff
@@ -854,6 +855,91 @@ def _unpack_epochs(epochs):
...
+        rank = len(this_picks)
+        if key == 'meg' and has_sss:
+            rank = min(rank, _get_rank_sss(info))
+        n_ssp = sum(pp['active'] for pp in info['projs'] if
+                    _match_proj_type(pp, this_info['ch_names']))
```
Comment (larsoner) [inline mne/cov.py]: I don't think it depends on `active` IIRC we always apply proj at some point in the pipeline before applying the inverse

File: mne/cov.py [hunk 5]
```diff
@@ -934,24 +1082,23 @@ def _compute_covariance_auto(data, method, info, method_params, cv,
                                    **method_params[this_method])
             sc.fit(data_)
             _info = None
-            estimator_cov_info.append((sc, sc.covariance_, _info))
-
-        elif this_method == 'pca':
+            estimator_cov_info.append([sc, sc.covariance_, _info])
+        elif this_method == 'pca':  # XXX this is now nonsensical
```
Comment (larsoner) [inline mne/cov.py]: what does this comment mean?

File: mne/cov.py [hunk 6]
```diff
@@ -934,24 +1082,23 @@ def _compute_covariance_auto(data, method, info, method_params, cv,
...
             pca, _info = _auto_low_rank_model(data_, this_method,
                                               n_jobs=n_jobs,
                                               method_params=mp, cv=cv,
                                               stop_early=stop_early)
             pca.fit(data_)
-            estimator_cov_info.append((pca, pca.get_covariance(), _info))
+            estimator_cov_info.append([pca, pca.get_covariance(), _info])
 
-        elif this_method == 'factor_analysis':
+        elif this_method == 'factor_analysis':  # XXX SSS case + EEG weird?
```
Comment (larsoner) [inline mne/cov.py]: and this one? is it exposed by some test? do we need a doc update?

File: mne/cov.py [hunk 7]
```diff
@@ -962,8 +1109,59 @@ def _compute_covariance_auto(data, method, info, method_params, cv,
...
-    for c in estimator_cov_info:
-        _undo_scaling_cov(c[1], picks_list, scalings)
+    for cov in estimator_cov_info:
+        # this is now a bit convoluted.
+        # The right part of the cov has to be picked, reprojected if needed
+        # and here the clipping has to happen.
+        # XXX we need a rank dict here and clip conditional upon
+        # low rank config.
+        if has_sss:
+            # XXX the logic is broken here for EEG
```
Comment (larsoner) [inline mne/cov.py]: This looks a bit like something we need to raise an error or warning for, or fix before release

File: mne/evoked.py [hunk 8]
```diff
@@ -358,7 +358,7 @@ def plot_field(self, surf_maps, time=None, time_label='t = %0.0f ms',
         return plot_evoked_field(self, surf_maps, time=time,
                                  time_label=time_label, n_jobs=n_jobs)
 
-    def plot_white(self, noise_cov, show=True):
+    def plot_white(self, noise_cov, rank=None, show=True):
```
Comment (larsoner) [inline mne/evoked.py]: should be after `show` for backward compat

File: mne/tests/test_cov.py [hunk 9]
```diff
@@ -588,7 +588,7 @@ def test_compute_covariance_auto_reg():
                         cov_b['data'][diag_mask]))
 
             # but the rest is the same
-            assert_array_equal(
+            assert_array_almost_equal(  # XXX make sure this is ok.
```
Comment (larsoner) [inline mne/tests/test_cov.py]: What does this comment mean?

File: tutorials/plot_compute_covariance.py [hunk 10]
```diff
@@ -35,26 +53,45 @@
...
+raw_empty_room.add_proj(
+    [pp.copy() for pp in raw.info['projs'] if 'EEG' not in pp['desc']])
+
+raw_empty_room = mne.preprocessing.maxwell_filter(
+    raw_empty_room, cross_talk=ctc_fname, calibration=fine_cal_fname,
+    coord_frame='meg')
+
+noise_cov = mne.compute_raw_covariance(
+    raw_empty_room, tmin=0, tmax=None, method='diagonal_fixed')
+noise_cov.nfree
```
Comment (larsoner) [inline tutorials/plot_compute_covariance.py]: leftover from debug?

File: tutorials/plot_compute_covariance.py [hunk 11]
```diff
@@ -35,26 +53,45 @@
...
+    bb for bb in raw.info['bads'] if 'EEG' not in bb]
+raw_empty_room.add_proj(
+    [pp.copy() for pp in raw.info['projs'] if 'EEG' not in pp['desc']])
+
+raw_empty_room = mne.preprocessing.maxwell_filter(
+    raw_empty_room, cross_talk=ctc_fname, calibration=fine_cal_fname,
+    coord_frame='meg')
+
+noise_cov = mne.compute_raw_covariance(
+    raw_empty_room, tmin=0, tmax=None, method='diagonal_fixed')
```
Comment (larsoner) [inline tutorials/plot_compute_covariance.py]: some explanation for `diagonal_fixed` choice would be good

File: tutorials/plot_compute_covariance.py [hunk 12]
```diff
@@ -35,26 +53,45 @@
...
+noise_cov.nfree
 ###############################################################################
-# Now that you the covariance matrix in a python object you can save it to a
-# file with :func:`mne.write_cov`. Later you can read it back to a python
-# object using :func:`mne.read_cov`.
+# Now that you the covariance matrix in an MNE-Python object you can save it
```
Comment (larsoner) [inline tutorials/plot_compute_covariance.py]: Now that you have

**Comment by jasmainak** [comment]
@dengemann I am sorry I need one more day before I can try it on biomag data ...

File: mne/cov.py (hunk 7 above)
Comment (dengemann) [reply mne/cov.py]: I should remove the XXX now....

File: mne/evoked.py (hunk 8 above)
Comment (dengemann) [reply mne/evoked.py]: Do we really support API positional arguments?
I feel something like show should be the last one. And people should anyways not use positional arguments. The order of the purely positional arguments should be of course preserved.

File: mne/tests/test_cov.py (hunk 9 above)
Comment (dengemann) [reply mne/tests/test_cov.py]: That assert_array_equal does not work here and that I don't know if this is expected or not.

File: tutorials/plot_compute_covariance.py (hunk 11 above)
Comment (dengemann) [reply tutorials/plot_compute_covariance.py]: it was just debugging. I will remove it.

File: tutorials/plot_compute_covariance.py (hunk 10 above)
Comment (dengemann) [reply tutorials/plot_compute_covariance.py]: yes

File: mne/evoked.py (hunk 8 above)
Comment (larsoner) [reply mne/evoked.py]: > Do we really support API positional arguments?

Yes, at least for the first few inputs (say maybe 10?). In this case there were only two so I think it's clear.

> I feel something like show should be the last one. 

We'd need to do this in a separate PR

>  And people should anyways not use positional arguments.

I think it's reasonable to use positional arguments and expect API stability there. Imagine someone has code:
```
show = True  # can be toggled to false in iteration

...
evoked.plot_white(cov, show)
```
This seems fine to me, and this change will break it.

File: mne/tests/test_cov.py (hunk 9 above)
Comment (larsoner) [reply mne/tests/test_cov.py]: I'd use `assert_allclose` with proper `rtol` and `atol` set based on the data of interest

File: mne/evoked.py (hunk 8 above)
Comment (dengemann) [reply mne/evoked.py]: I don't share this view as I always recommend to pass by keyword, but if you insist it's ok with me.

Comment (larsoner) [reply mne/evoked.py]: Yes please

**Comment by agramfort** [comment]
@dengemann you have some undesired commits in this PR (stuff already in master)

**Comment by agramfort** [comment]
@dengemann you need to rebase and clean history...

**Comment by dengemann** [comment]
Copied that. Did not have time yet to fix the mess.

**Comment by larsoner** [comment]
Travis failure is the unrelated OSX failure currently in `master`.

Can you add to `whats_new.rst` to summarize the change? Maybe toward the top of BUG so it's clear what's fixed. Otherwise LGTM!

File: mne/cov.py (hunk 4 above)
Comment (dengemann) [reply mne/cov.py]: what do you mean @larsoner? This is what I needed to get it right IIRC. Any alternative suggestion?

**Comment by cbrnr** [comment]
Slightly OT, but do you think this can be merged ASAP to finalize 0.15 (if possible, please consider #4662, which is ready to go, but I'd argue for no further major delays)?

File: mne/cov.py (hunk 4 above)
Comment (larsoner) [reply mne/cov.py]: I'm trying to figure out if you should just do `sum(_match_proj_type(pp, this_info['ch_names']))`. Think about the `proj='delayed'` use case or even `epochs=Epochs(proj=False)` (but will be applied later during inversion) will this code do the right thing?

**Comment by larsoner** [comment]
> but do you think this can be merged ASAP to finalize 0.15

+1 yes I think this is all we are waiting for currently

File: tutorials/plot_compute_covariance.py [hunk 13]
```diff
@@ -16,15 +16,21 @@
...
+# construct a noise covariance matrix that can be used when computing the
+# inverse solution. For more information, see :ref:`BABDEEEB`.
 data_path = sample.data_path()
 raw_empty_room_fname = op.join(
     data_path, 'MEG', 'sample', 'ernoise_raw.fif')
 raw_empty_room = mne.io.read_raw_fif(raw_empty_room_fname)
 raw_fname = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw.fif')
 raw = mne.io.read_raw_fif(raw_fname)
-raw.set_eeg_reference('average', projection=True)
+raw.set_eeg_reference()
```
Comment (larsoner) [inline tutorials/plot_compute_covariance.py]: This should be reverted

File: mne/cov.py (hunk 4 above)
Comment (agramfort) [reply mne/cov.py]: I agree with @larsoner we always end up applying all proj in the inverse to me:

sum(_match_proj_type(pp, this_info['ch_names']))

seems like the right fix

**Comment by dengemann** [comment]
Closing to split into mergeable and more-dev-to-do part.

**Pull request closed without merging**